scalalib/src/scoverage/ScoverageModule.scala

```diff
@@ -0,0 +1,48 @@
+package mill.lib.scoverage
+
+import mill._
+import mill.scalalib._
+
+import _root_.scoverage.Serializer.{ coverageFile, deserialize }
+import _root_.scoverage.IOUtils.{ findMeasurementFiles, invoked }
+import _root_.scoverage.report.ScoverageHtmlWriter
+
+trait ScoverageModule { outer: ScalaModule =>
+
+  trait ScoverageCompile extends ScalaModule {
+    def dataDir: String
+    def sources = outer.sources
+    def resources = outer.resources
+    def scalaVersion = outer.scalaVersion()
+    def compileIvyDeps = outer.compileIvyDeps()
+    def ivyDeps = outer.ivyDeps() ++
+      Agg(ivy"org.scoverage::scalac-scoverage-runtime:1.3.1")
+    def scalacPluginIvyDeps = outer.scalacPluginIvyDeps() ++
+      Agg(ivy"org.scoverage::scalac-scoverage-plugin:1.3.1")
+    def scalacOptions = outer.scalacOptions() ++
+      Seq(s"-P:scoverage:dataDir:$dataDir")
+
+    def html() = T.command {
+      val coverageFileObj = coverageFile(dataDir)
+      if (coverageFileObj.exists) {
+        val coverage = deserialize(coverageFileObj)
+        coverage(invoked(findMeasurementFiles(dataDir)))
+        val Seq(PathRef(sourceFolderPath, _, _)) = sources()
+        val sourceFolders = Seq(sourceFolderPath.toIO)
+        val htmlFolder = new java.io.File(s"${dataDir}/html")
+        htmlFolder.mkdir()
+        new ScoverageHtmlWriter(sourceFolders, htmlFolder, None)
+          .write(coverage)
+      } else {
+        T.ctx().log.error(s"Cannot write scoverage report. Directory ${dataDir} does not exist!")
```

Comment:
That means what? What should I do, when I see this error message? Who is supposed to create the data dir for me? Does that mean, scoverage failed?

Reply:
`dataDir` is where scoverage will output it's measurement files + also any reports it generates.

After reading more about the Task context api, I think it may be better to use the allocated `out/foo/scoverage/...` directory tree for a `object foo extends ScalaModule` to store the measurements and reports.

Is there a way to access a `out/foo` from within the body of the `ScalaModule` object? I know I can access `T.ctx().dest` from within a task/command, but this seems different.

Reply:
I believe I can use anonymous tasks actually to handle this issue. I'll give it a go